src/extension_traits/vec.rs

```diff
@@ -0,0 +1,346 @@
+    ///
+    /// let mut extra: Out<'_, [u8]> = vec.reserve_uninit(WORLD.len());
+    /// extra.r().copy_from_slice(WORLD);
+    ///
+    /// // `.reserve_uninit()` guarantees the following properties:
+    /// assert_eq!(extra.len(), WORLD.len());
+    /// let extra_start: *mut u8 = extra.r().as_mut_ptr().cast();
+    /// let uninit_start: *mut u8 = vec.as_mut_ptr().wrapping_add(vec.len());
+    /// assert_eq!(extra_start, uninit_start);
+    ///
+    /// unsafe {
+    ///     // # Safety
+    ///     //
+    ///     //   - `.copy_from_slice()` contract guarantees initialization
+    ///     //     of `extra`, which, in turn, from `reserve_uninit`'s contract,
+    ///     //     leads to the `vec` extra capacity having been initialized.
+    ///     vec.set_len(vec.len() + WORLD.len());
+    /// }
+    /// assert_eq!(
+    ///     vec,
+    ///     b"Hello, World!",
+    /// );
+    /// ```
+    #[inline]
+    fn reserve_uninit (self: &'_ mut Vec<T>, additional: usize)
+      -> Out<'_, [T]>
+    {
+        self.reserve(additional);
+        let (_, extra) = self.split_at_extra_cap();
+        unsafe {
+            // Safety: `Vec<T>` guarantees that `cap >= len + additional` and
+            // thus that `cap - len >= additional`.
+            extra.get_out_unchecked(.. additional)
+        }
+    }
+
+    /// Gets an [`&out [T]`][`Out`] slice (of `self.capacity()` elements)
+    /// to the backing buffer.
+    #[inline]
+    fn get_backing_buffer (self: &'_ mut Vec<T>)
+      -> Out<'_, [T]>
+    where
+        T : Copy, // Opinionated stance against accidental memory leaks
+    {
+        self.get_backing_buffer_manually_drop()
+    }
+
+    /// Same as [`.get_backing_buffer()`][`VecCapacity::get_backing_buffer`]
+    /// but without the `Copy` bound.
+    ///
+    /// This means that extra care should be taken if
+    /// `mem::needs_drop::<Self::Item>()`. Indeed, if the `OutSlice` is used to
+    /// overwrite initialized elements, then such elements will be destroyed
+    /// without their `.drop()` glue / destructors ever being run.
+    ///
+    /// ⚠️ **Misusage of this function can thus lead to memory leaks** ⚠️
+    ///
+    /// # Counter-example
+    ///
+    /// ```rust
+    /// use ::uninit::prelude::*;
+    ///
+    /// let mut v = vec![Some(Box::new(42))];
+    /// v   .get_backing_buffer_manually_drop()
+    ///     .get_out(0)
+    ///     .unwrap()
+    ///     .write(None) // the box is not freed despite now being unreachable.
+    /// ;
+    /// ```
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use ::uninit::prelude::*;
+    /// use ::std::cell::Cell;
+    ///
+    /// let mut v = vec![Cell::new(0)];
+    /// v   .get_backing_buffer_manually_drop() // No drop glue, so this is fine
+    ///     .get_out(0)
+    ///     .unwrap()
+    ///     .write(Cell::new(42))
+    /// ;
+    /// assert_eq!(v[0].get(), 42);
+    /// ```
+    #[inline]
+    fn get_backing_buffer_manually_drop (self: &'_ mut Vec<T>)
```

Comment:
Yeah, I struggle to find an acceptable way to protect from accidental leaks 🤔 Two questions:

  - how would you design it? I find that we should not be able to write stuff like `let mut s: String = ...; s.as_out().write("...".into())`, hence the `s.as_manually_drop_mut().as_out()` which makes it clearer, but for `Vec` I do not know how to do it. One way way would be to get rid of the `T : Copy` bounds on the methods and implement the trait for both `Vec<impl Copy>` and `Vec<ManuallyDrop<_>>`, with a conversion helper between a classic `Vec` and the latter.

  - > that would be unsound
    
    Could you elaborate? If I am to take that second road, I have to be careful that it does not lead to unsoundness. From my understanding, contrary to `MaybeUninit`, casting back and forth between `T` and `ManuallyDrop<T>` is always sound, since the `ManuallyDrop` wrapper is safe to both construct and deconstruct. So, for instance, someone doing

    ```rust
    ManuallyDrop::new({
        struct UBOnDrop; impl Drop for UBOnDrop {
            fn drop(&mut self) { unsafe { hint::unreachable() } }
        }
        UBOnDrop
    })
    ```

    can only do so soundly if they "have control" / knowledge of the exact usage of that value (in other words, their codebase always has ownership of it), so that relinquishing ownership of that value into some black-box API (_e.g._, `some_collection.push()`) is actually unsound, since that API would be allowed to `ManuallyDrop::into_inner()`.

Reply:
>  One doesn't get a slice of ManuallyDrop<T> and even more, that would be unsound.

We can't give the whole backing buffer as `[ManuallyDrop<T>]` as not all of it is initialized. And conversely, you can't return the whole buffer as `[MaybeUninit<T>]` as some of it *must stay* initialized. So the `&out ` option is in fact the only sound choice.

Reply:
> how would you design it? 

I would choose a name connected to `std::mem::forget` or `Box::leak`. Both of those are established safe, but leaking, operations. What do you think of `as_forgetting_out`?

Reply:
> We can't give the whole backing buffer as [ManuallyDrop<T>] as not all of it is initialized

Oh that for sure! I was thinking that you meant that something like `&mut T <-> &mut ManuallyDrop<T>` would be unsound, and I just couldn't see it